src/mlpack/methods/matrix_completion/matrix_completion.hpp

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file matrix_completion.hpp
+ * @author Stephen Tu
+ *
+ * A thin wrapper around nuclear norm minimization to solve
+ * low rank matrix completion problems.
+ */
+#ifndef __MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_HPP
+#define __MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_HPP
+
+#include <mlpack/core/optimizers/lrsdp/lrsdp.hpp>
+
+namespace mlpack {
+namespace matrix_completion {
+
+/**
+ * This class implements the popular nuclear norm minimization heuristic for
+ * matrix completion problems. That is, given known values M_ij's, the
+ * following optimization problem (semi-definite program) is solved to fill in
+ * the remaining unknown values of X
+ *
+ *   min ||X||_* subj to X_ij = M_ij
+ *
+ * where ||X||_* denotes the nuclear norm (sum of singular values of X).
+ *
+ * For a theoretical treatment of the conditions necessary for exact recovery,
+ * see the following paper:
+ *
+ *   A Simpler Appoarch to Matrix Completion.
+ *   Benjamin Recht. JMLR 11.
+ *   http://arxiv.org/pdf/0910.0651v2.pdf
+ *
+ * An example of how to use this class is shown below:
+ *
+ * @code
+ * size_t m, n;        // size of unknown matrix
+ * arma::umat indices; // contains the known indices [2 x n_entries]
+ * arma::vec values;   // contains the known values [n_entries]
+ *
+ * MatrixCompletion mc(m, n, indices, values);
+ * mc.Recover();
+ * mc.Recovered();     // access completed matrix
+ * @endcode
+ *
+ * @see LRSDP
+ */
+class MatrixCompletion
+{
+ public:
+
+  /**
+   * Construct a matrix completion problem, specifying the maximum rank of the
+   * solution.
+   *
+   * @param m Number of rows of original matrix.
+   * @param n Number of columns of original matrix.
+   * @param indices Matrix containing the indices of the known entries (must be
+   *    [2 x p]).
+   * @param values Vector containing the values of the known entries (must be
+   *    length p).
+   * @param r Maximum rank of solution.
+   */
+  MatrixCompletion(const size_t m,
+                   const size_t n,
+                   const arma::umat& indices,
+                   const arma::vec& values,
+                   const size_t r);
```

Comment:
I like the idea of `arma::sp_mat`. However, it does leave open the following ambiguity: what if i want a particular entry (i, j) to be zero? How do i distinguish this from the case where I don't know the value of entry (i, j). The only way seems to be to force us to break the abstraction of a sparse matrix. Sure, I could say something like `X(i, j) = 0` explicitly, but what if the particular setter implementation does `if (value == 0) return;`?

Alternatively, you could have the semantic that the `sp_mat` constructor doesn't the constraint that a particular entry is zero, but then you push the burden onto API callers to determine if this is the case for their data.